---
title: "Giotto Processing"
output: github_document
---
# Process Visium Transcriptomics Data with Giotto


## Initialize Giotto and Read Necessary Files
1. Read in Count UMI Files
2. Read in Spatial information (Tissue locations, metadata)
3. Create Giotto Object
```{r}
data_path = '.'
workdir = 'workdir'
myinst=createGiottoInstructions(save_plot = T,
                                show_plot = T,
                                return_plot = F,
                                save_dir = workdir)

expr <- data.table::fread('../spatial/GV25999_human_counts.tsv.gz',sep = '\t',header = T) %>% as.data.frame()
rownames(expr) <- expr$GENE
expr <- expr[,2:ncol(expr)]

#Read in Tissue Information
spatial_dir <- 'spatial'
spatial_results <- read.table('../data/spatial/tissue_positions_list.csv',sep = ',')
colnames(spatial_results) <- c("cell_ID", "in_tissue", "array_row", "array_col", "col_pxl", "row_pxl")
spatial_results$cell_ID <- gsub('-1','',spatial_results$cell_ID)
spatial_results <- spatial_results[match(colnames(expr),spatial_results$cell_ID),]
spatial_locs <- spatial_results[,c('row_pxl','col_pxl')]
spatial_locs$col_pxl <- -spatial_locs$col_pxl
colnames(spatial_locs) <- c('sdimx','sdimy')

#create Giotto object
vehicle <- createGiottoObject(expression= expr,spatial_locs = spatial_locs,cell_metadata = spatial_results,
                              instructions = myinst)

```


## Subset Giotto Object and Filter Low Quality Cells
1. Plot In Tissue Spots (detected tissue from 10x workflow)
2. View metadata
3. Normalize, Filter, and Determine Covariate Features
4. Plot Filtered Dataset Spatially
```{r}
spatPlot(gobject= vehicle,point_size =2, 
         cell_color='in_tissue',
         cell_color_code = c('0'='lightgrey','1'='blue'),
         save_param=c(save_name='1-spatplot'))

metadata=pDataDT(vehicle)


#normalize
vehicle <- processGiotto(vehicle,
                         filter_params = list(expression_threshold = 1,
                                              feat_det_in_min_cells = 25,
                                              min_det_feats_per_cell = 250),
                         norm_params = list(norm_methods = 'standard',
                                            scale_feats = TRUE,
                                            scalefactor = 6000),
                         stat_params = list(expression_values = 'normalized'),
                         adjust_params = list(expression_values = c('normalized'),
                                              covariate_columns = 'nr_feats'))

#location of spots
spatPlot(gobject=vehicle, point_size =2.5, save_param=c(save_name='2-spatplot'))
spatPlot(gobject=vehicle,cell_color='nr_genes',color_as_factor=F,point_size=2.5,
         save_param=c(save_name='3-spatplot'))
```

## Visualize QC Metrics Post Filtering
1. Show spatial plot
2. Plot Distribution of counts
3. Plot Number of Genes Per Spot
4. Plot Number of Spots Each Gene is Present in
```{r}
spatPlot(gobject=vehicle , point_size =2, save_param=c(save_name='2-spatplot'))
nr_genes <- spatPlot(gobject=vehicle ,cell_color='nr_genes',color_as_factor=F,point_size=2,
         save_param=c(save_name='3-spatplot'),return_plot =T)

#Run QC 
all_counts <- as.data.frame(unlist(as.list(vehicle@norm_expr)))
colnames(all_counts) <- c('counts')
count_histogram <- ggplot(all_counts,aes(counts))+ geom_histogram(binwidth = 0.1)+xlab('Normalized Count') + ylab('Count') + ggtitle('vehicle_xenomake') + theme(plot.title = element_text(hjust = 0.5))
cell_metadata <- vehicle@cell_metadata
cell_metadata$sample <- 'vehicle_xenomake'
p1 <- ggplot(cell_metadata,aes(x=sample,y=nr_genes,color = sample)) + geom_boxplot() + ggtitle('number of Genes per Spot') + xlab(NULL) + ylab('nr_genes')+theme_classic()+theme(plot.title = element_text(hjust=0.5))
gene_metadata <- vehicle@gene_metadata
gene_metadata$sample <- 'vehicle_xenomake'
p2 <- ggplot(gene_metadata,aes(x=sample,y=nr_cells,color=sample))+ geom_boxplot() + ggtitle('Number of Spots per Gene') + xlab(NULL) + ylab('nr_cell')+theme_classic()+theme(plot.title = element_text(hjust=0.5))

library(ggpubr)
p3 <- ggarrange(p1,p2,common.legend = TRUE,legend = 'bottom')
p4 <- ggarrange(nr_genes,filter_combinations)
ggarrange(p3,p4,count_histogram,ncol=1,nrow = 3) -> vehicle_qc

```

## Calculate Highly Variable Genes Based on Normalized Expression 
```{r}
vehicle <- calculateHVF(gobject = vehicle, method = 'cov_groups')
```

## Perform Dimensionality Reduction using PCA
```{r}
feature_metadata = getFeatureMetadata(vehicle)[]
featgenes = feature_metadata[hvf == 'yes' & perc_cells > 4 & mean_expr_det > 0.5]$feat_ID
vehicle <- runPCA(gobject = vehicle, feats_to_use = featgenes, scale_unit = F, center = F)

plotPCA(gobject = vehicle)
```

## Plot Reduced Dimensions using UMAP and TSNE
```{r}
vehicle <- runUMAP(vehicle, dimensions_to_use = 1:30)
plotUMAP(gobject = vehicle)

```

## Create Network
```{r}
vehicle <- createNearestNetwork(gobject = vehicle, type = "sNN", dimensions_to_use = 1:30, k = 15)
```


## Cluster Cells and Show clusters on UMAP
```{r}
## k-means clustering
vehicle <- doKmeans(gobject = vehicle, dim_reduction_to_use = 'pca')

## Leiden clustering - increase the resolution to increase the number of clusters
vehicle <- doLeidenCluster(gobject = vehicle,
                            resolution = 0.25,
                            n_iterations = 1000,
                            name = 'leiden_0.4_1000')

## Louvain clustering - increase the resolution to increase the number of clusters
# The version argument may be changed to 'multinet' to run a Louvain algorithm
# from the multinet package in R.
vehicle <- doLouvainCluster(gobject = vehicle,
                            version = 'community',
                            resolution = 0.4)

#Plot UMAP post-clustering to visualize Leiden clusters
plotUMAP(gobject = vehicle,
         cell_color = 'leiden_0.4_1000',
         show_NN_network = T,
         point_size = 2.5)
```

# Perform Spatially Aware clustering


## Spatial Network Analysis
```{r}
vehicle <- createSpatialNetwork(gobject = vehicle,minimum_k = 2,name = 'spatial_network')
spatPlot2D(gobject = vehicle,
           show_network= T,
           network_color = 'blue')
```

## Perform Binarization of Counts
```{r}
## rank binarization
ranktest = binSpect(vehicle, bin_method = 'rank',
                    calc_hub = T, hub_min_int = 5,
                    spatial_network_name = 'spatial_network',return_gobject = TRUE)

```

## Plot Spatial Gene Expression For Top Spatially Localized Genes
```{r}
spatFeatPlot2D(vehicle,
               expression_values = 'scaled',
               feats = ranktest$feats[1:6], cow_n_col = 2, point_size = 1.5)
```


## Detect Spatial Co-Expression Patterns
```{r}
# 3.1 cluster the top 500 spatial genes into 20 clusters
ext_spatial_genes = ranktest[1:1500,]$feats

# here we use existing detectSpatialCorGenes function to calculate pairwise distances between genes (but set network_smoothing=0 to use default clustering)
spat_cor_netw_DT = detectSpatialCorFeats(vehicle,
                                         method = 'network',
                                         spatial_network_name = 'spatial_network',
                                         subset_feats = ext_spatial_genes)


# 3.3 identify potenial spatial co-expression
spat_cor_netw_DT = clusterSpatialCorFeats(spat_cor_netw_DT, name = 'spat_netw_clus', k = 8)

# visualize clusters
heatmSpatialCorFeats(vehicle,
                     spatCorObject = spat_cor_netw_DT,
                     use_clus_name = 'spat_netw_clus',
                     heatmap_legend_param = list(title = NULL),
                     save_param = list(base_height = 6, base_width = 8, units = 'cm'))

# 3.4 create metagenes / co-expression modules
cluster_genes = getBalancedSpatCoexpressionFeats(spat_cor_netw_DT, maximum = 30)
vehicle = createMetafeats(vehicle, feat_clusters = cluster_genes, name = 'cluster_metagene')


# 3.2 identify most similar spatially correlated genes for one gene
spatFeatPlot2D(vehicle,
               expression_values = 'scaled',
               feats = top10_genes$variable[1:4], point_size = 3)


```


### Perform Spatially informed clustering
```{r}
my_spatial_genes = names(cluster_genes)

vehicle <- runPCA(gobject = vehicle,
                      feats_to_use = my_spatial_genes,
                      name = 'custom_pca')

vehicle <- runUMAP(vehicle,
                       dim_reduction_name = 'custom_pca',
                       dimensions_to_use = 1:20,
                       name = 'custom_umap')

vehicle <- createNearestNetwork(gobject = vehicle,
                                    dim_reduction_name = 'custom_pca',
                                    dimensions_to_use = 1:20, k = 5,
                                    name = 'custom_NN')

vehicle <- doLeidenCluster(gobject = vehicle,
                               network_name = 'custom_NN',
                               resolution = 0.1, n_iterations = 1000,
                               name = 'custom_leiden')

spatPlot2D(vehicle,
           cell_color = 'custom_leiden', point_size = 4)
```


